### ISSUE TITLE
text/html on clipboard clobbered from Linux server to Windows client

### ISSUE THREAD

**Issue by jtbraun** [synergy/synergy#5166]
Repro steps:
* Select some text in a web browser, right click and copy (gets a text/html atom on the clipboard)
* Verify that by running `for i in primary secondary buffer clipboard; do echo $i; xclip -selection $i -o -t TARGETS; echo; done`
* move to the windows client, paste into an HTML-aware application (Outlook message editor, Word, etc)
** Note that the converted plain text is pasted, not the HTML-ish version

Here's the DEBUG1 log from the host at the time I switched from Linux Host-> Windows Client screens:
[SynergyClobberHtml_hostlog.txt](https://github.com/synergy/synergy/files/73630/SynergyClobberHtml_hostlog.txt)
And here's a summary of the CF_* contents of the clipboard on the Windows Client afterward:
[SynergyClobberHtml_clientclipboard.txt](https://github.com/synergy/synergy/files/73629/SynergyClobberHtml_clientclipboard.txt)

The formats available just after the copy and before the screen switch:
```
$ for i in primary secondary buffer clipboard; do echo $i; xclip -selection $i -o -t TARGETS; echo; done
primary
TARGETS
SAVE_TARGETS
MULTIPLE
STRING
UTF8_STRING
TEXT
text/plain

secondary
TARGETS
UTF8_STRING

buffer
synergy/synergy
clipboard
TARGETS
SAVE_TARGETS
MULTIPLE
STRING
UTF8_STRING
TEXT
text/plain
text/html
```

It looks like the bug is on the client.  Both clipboards are received (clipboard 0 with the larger size must contain the text/html data of interest), and clipboard 0 puts format:2 (kHTML) onto the windows clipboard.  However, `IClipboard::unmarshall` appears to empty the clipboard on every call.  So while the data is received, only host clipboard 1's data ever actually appears in the windows client, once all the unmarshalling finishes.
```
[2015-12-28T16:03:06] INFO: entering screen
[2015-12-28T16:03:06] DEBUG: start receiving clipboard data
[2015-12-28T16:03:06] DEBUG: receiving clipboard 0 size=2204
[2015-12-28T16:03:06] DEBUG: received clipboard 0 size=2204
[2015-12-28T16:03:06] DEBUG: open clipboard
[2015-12-28T16:03:06] DEBUG: empty clipboard
[2015-12-28T16:03:06] DEBUG: add 15 bytes to clipboard format: 0
[2015-12-28T16:03:06] DEBUG: add 2169 bytes to clipboard format: 2
[2015-12-28T16:03:06] DEBUG: close clipboard
[2015-12-28T16:03:06] INFO: clipboard was updated
[2015-12-28T16:03:06] DEBUG: start receiving clipboard data
[2015-12-28T16:03:06] DEBUG: receiving clipboard 1 size=27
[2015-12-28T16:03:06] DEBUG: received clipboard 1 size=27
[2015-12-28T16:03:06] DEBUG: open clipboard
[2015-12-28T16:03:06] DEBUG: empty clipboard
[2015-12-28T16:03:06] DEBUG: add 15 bytes to clipboard format: 0
[2015-12-28T16:03:06] DEBUG: close clipboard
[2015-12-28T16:03:06] INFO: clipboard was updated
```

**Comment by XinyuHou** [comment]
@jtbraun 
Is this duplicated to another issue you raised? #5165

**Comment by jtbraun** [comment]
No, it's not a duplicate.  #5165 is a bug/problem on the Linux X11 Clipboard host.  It reads the "UTF8_STRING" atom and treats it as a UTF16-encoded "text/html" atom and marshals it across to the client.

This is a separate issue on the Windows client.  The windows client receives two different clipboards, with different contents, but doesn't do a great job of resolving the two transferred X11 clipboards down to a single clipboard on the windows client.  As to how to deal with this, likely the Synergy host should probably keep track of which clipboard number was most recently modified on the linux host, and the windows client, when it receives multiple clipboards, should choose the most-recently modified one.  Alternatively, you could have the windows client try to merge multiple clipboards into the single on or something, but how you determine which one "wins" may require a new user preference (prefer SELECTION over CLIPBOARD, or something).

In any case, these are different issues, not duplicates.

**Comment by XinyuHou** [comment]
Maybe related to #2909

**Comment by XinyuHou** [comment]
@jtbraun 
What Linux flavor do you use?

I tried to reproduce this on a Ubuntu 15.04 server and a Windows 7 client.

Client receive format 0 and 1 for both clipboard 0 and 1. I think the problem is for some reasons your server doesn't send format 1 in clipboard 1.

**Comment by jtbraun** [comment]
Interesting.  I'm using Ubuntu 14.04 LTS for the server and Windows 7 64-bit on the client. I ran some experiments and have some client logs attached as 
[CopyPasteMulticlipboardsLog.txt](https://github.com/synergy/synergy/files/77784/CopyPasteMulticlipboardsLog.txt)

The scripts clearclip.sh and printclip.sh are in 
[xcliphelpers.zip](https://github.com/synergy/synergy/files/77786/xcliphelpers.zip)

A summary of the actions taken on the host.  After each action I went to the client, pasted, and moved back to the host:
* Clear the clipboard (with clearclip.sh)
** Result: blank clipboard on the client (expected).  zero-length data for formats 0,1,2 processed on the client.
* Perform a selection (with the mouse) in chromium-browser, but do NOT copy
** Result: the "primary" host clipboard has text but not HTML, that's probably a chromium bug.  The text pastes correctly in the client, only format 0 from clipboard 1 was processed on the client.
* Copy the previously-made selection in chromium-browser (without reselecting)
** Result: mojibake (result of bug #5165). Formats 0 and 2 were processed from clipboard 0
* Clear the clipboard (with clearclip.sh)
** Result: blank clipboard on the client (expected).  zero-length data for formats 0,1,2 processed on the client.
* Select AND Copy in chromium-browser
** Result: clipboard on client contains the correct text data, but no longer contains any html data (this bug).
*** Clipboard 0 was processed on the client, and produced formats 0 and 2.
*** Then Clipboard 1 was processed on the client, and produced format 0 only.


So, depending on the order of select/copy on the host and when I move the mouse to the client, I can influence if only clipboard 0 or clipboard 1 is sent to the client.  In the case where BOTH clipboards are sent (the final one), the contents of clipboard 1 are the only ones that remain on the client's clipboard, the format 2 data from clipboard 0 is removed by the empty clipboard call at the start of processing clipboard 1.

**Comment by XinyuHou** [comment]
@jtbraun 

Thank you for the detail information.

I did another test.

Only selection copy both plain text and HTML to client on my side. Select and Copy does the same.

The difference apart from the OS is I was using firefox instead of chromium.

Do you have firefox on you machine?

**Comment by jtbraun** [comment]
I do have firefox.  It looks like firefox (unlike chromium) places text/html on both the X11 "primary" on selection and "clipboard" on copy.

And I can confirm that selection, copy, and select+copy (as detailed above) actually produces a "format 2" entry on the windows client that pastes correctly with the correct encoding, etc.  

So maybe it's chromium that needs to change to also insert text/html into the X11 "primary" clipboard to "fix" this, unless you guys want to implement a workaround specifically for chromium-browser.

Regarding the encoding in #5165, I'll update it to note that the behavior is different between firefox and chromium-browser, though the bug is still in synergy, treating the UTF8_STRING atom as being UTF16-encoded as noted.

**Comment by XinyuHou** [comment]
@jtbraun 

Thanks for all the help.

I think it's a bug in Synergy or at least something we can improve. Those 2 clipboards should work more independently.

**Comment by jtbraun** [comment]
Ugh.  Here's an article from someone who also was trying to work out how html is put onto the X11 clipboard:
https://hsivonen.fi/kesakoodi/clipboard/

Gecko (under firefox) actually produces four related values:
```
text/html
text/_moz_htmlcontext
text/_mox_htmlinfo
text/x-moz-url-prov
```
It seems that these serve similar purposes to the start/end fragment stuff for the windows clipboard.  These are always UTF16-hostorder encoded.  (also all briefly mentioned here https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsITransferable)

Chromium (a derivative of webkit) seems to provide a meta tag at the start to indicate the encoding, without any context (or its translated the context), all in a text/html atom:
```html
 <meta http-equiv="content-type" content="text/html; charset=utf-8"><span style="color: rgb(51, 51, 51); font-family: 'Helvetica Neue', Helvetica, 'Segoe UI', Arial, freesans, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'; font-size: 14px; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: 22.4px; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 1; word-spacing: 0px; -webkit-text-stroke-width: 0px; display: inline !important; float: none; background-color: rgb(255, 255, 255);">I think it's a bug in Synergy or at least something we can improve. Those 2 clipboards should work more independently.</span>
```

I can't find a spec that details that, you'd probably have to look at webkit's code.

If you REALLY wanted to support transferring HTML, you'd survey a bunch of browsers, and implement each of their separate behaviors.  It's gross.
